src/main/java/seedu/address/logic/commands/ListCommand.java

```diff
@@ -1,24 +1,92 @@
 package seedu.address.logic.commands;
 
-import static java.util.Objects.requireNonNull;
-import static seedu.address.model.Model.PREDICATE_SHOW_ALL_PERSONS;
+import java.util.Optional;
+import java.util.function.Predicate;
 
+import seedu.address.logic.commands.exceptions.CommandException;
 import seedu.address.model.Model;
+import seedu.address.model.person.Address;
+import seedu.address.model.person.Person;
+import seedu.address.model.tag.Tag;
 
 /**
  * Lists all persons in the address book to the user.
  */
 public class ListCommand extends Command {
 
+    private final Optional<Address> address;
+    private final Optional<String> category;
+    private final Optional<Gender> gender;
+    private final Optional<Tag> tag;
+
+    public static final String MESSAGE_ARGUMENTS = "ADDRESS: %s, CATEGORY: %s, GENDER: %s, TAG: %s";
+
     public static final String COMMAND_WORD = "list";
 
-    public static final String MESSAGE_SUCCESS = "Listed all persons";
+    public static final String MESSAGE_SUCCESS = "Listed all persons with specifications: " + MESSAGE_ARGUMENTS;
 
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Lists all enrolled users who fit the specified criteria, "
+            + "or all enrolled users if no criteria were specified.\n"
+            + "Parameters: \n"
+            + "<optional> c/ [CATEGORY]\n"
+            + "<optional> g/ [GENDER]\n"
+            + "<optional> a/ [ADDRESS]\n"
+            + "<optional> t/ [TAG]\n"
+            + "Example: " + COMMAND_WORD + "  "
+            + "c/ nurse";
+
+    /**
+     * @param a address to be filtered
+     * @param c category (nurse/patient) to be filtered
+     * @param g gender to be filtered
+     * @param t tag to be filtered
+     */
+    public ListCommand(Optional<Address> a, Optional<String> c, Optional<Gender> g, Optional<Tag> t) {
+        address = a;
+        category = c;
+        gender = g;
+        tag = t;
+    }
 
     @Override
     public CommandResult execute(Model model) {
-        requireNonNull(model);
-        model.updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
-        return new CommandResult(MESSAGE_SUCCESS);
+        Predicate<Person> predicate = x -> {
+            boolean addressMatch = x.getAddress().equals(address.orElse(x.getAddress()));
+            boolean categoryMatch = x.getCategory().equals(category.orElse(x.getCategory()));
+            boolean genderMatch = x.getGender().equals(gender.orElse(x.getGender()));
+
+            Predicate<Tag> tagPredicate = y -> {
+                Tag tagToCompare = tag.orElse((Tag) x.getTags().toArray()[0]);
```

Comment:
I think there's a better way to write this, but I haven't thought of it yet.

Reply:
I think for now its okay, maybe we can split up the predicates and then combine the predicates.